### PR TITLE
Wait for beacon node to get online when validator starts

### LINKED ISSUE #8188: Validator Client crash-loops until beacon is accessible and sync'd (to beacon chain?)
# 🚀 Feature Request

### Description

When we deploy a new eth2 cluster (redundant beacons and multiple clients) to our platform, the beacon node takes some time for its DNS name to resolve; until this resolves, the client crash-loops failing to connect.
Once it can connect, the client crash-loops for several minutes more until the beacon node is sync'd (presumably to the eth2 beacon chain).  This crash-looping may be the intended behavior but it's not user-friendly or platform-friendly.

e.g. after connection made:
```
time="2021-01-02 21:05:34" level=info msg="Waiting for beacon chain start log from the ETH 1.0 deposit contract" prefix=validator
time="2021-01-02 21:10:36" level=fatal msg="Could not determine if beacon chain started: could not receive ChainStart from stream: rpc error: code = Unavailable desc = transport is closing" prefix=validator
```

When it finally stops crashing:
```
time="2021-01-02 21:15:36" level=info msg="Waiting for beacon chain start log from the ETH 1.0 deposit contract" prefix=validator
time="2021-01-02 21:15:36" level=info msg="Beacon chain started" genesisTime=2020-12-01 12:00:23 +0000 UTC prefix=validator
```

### Describe the solution you'd like

A client should be able to be started before its beacon node and be resilient and patient.  If nothing else, this behavior should be optional (`--no-fail-fast`).

### Describe alternatives you've considered

If we can't get this added, we'll need to build our own readiness probes to hold off starting the client until it's likely not going to crash.  Pod crash-loops are something our platforms alerts on as, in general, crashing under 'normal operation' should be avoided.

### PR DESCRIPTION
**Pull request by shayzluf**
**What type of PR is this?**

Feature

**What does this PR do? Why is it needed?**
When starting the validator client together or before running the beacon node validator crashes as wait for chin start stream is not being initiated as validator client can't connect with the beacon node.
This pr polls the endpoint of beacon node while it is nor reachable and waits till its up to get on with its functionality. the client polling flow has been taken from an already used flow in slasher though the endpoint that is being polled is different.

**Which issues(s) does this PR fix?**

Fixes #8188

### CODE REVIEW

**Review by prestonvanloon — CHANGES_REQUESTED** [review]
Please add tests for this new functionality.

File: validator/client/validator.go [hunk 1]
```diff
@@ -176,6 +179,32 @@ func (v *validator) WaitForChainStart(ctx context.Context) error {
...
 }
 
+// Poll the beacon node every chainStartPollingInterval until the node
+// is available.
+func (v *validator) queryChainStart(ctx context.Context) (ethpb.BeaconNodeValidator_WaitForChainStartClient, error) {
+	stream, err := v.validatorClient.WaitForChainStart(ctx, &ptypes.Empty{})
+	if err == nil {
+		return stream, nil
+	}
+	log.WithError(err).Error("Could not fetch chain start from beacon node")
```
Comment (terencechain) [inline validator/client/validator.go:189]: Wouldn't we always be logging `Could not fetch chain start from beacon node` here?

**Comment by shayzluf** [comment]
@prestonvanloon added test and fixed old tests

File: validator/client/validator.go (hunk 1 above)
Comment (shayzluf) [reply validator/client/validator.go]: could you clarify. i changed the logging to warning wdyt?

File: validator/client/validator.go [hunk 2]
```diff
@@ -44,6 +44,10 @@ var reconnectPeriod = 5 * time.Second
 // in case no keys were fetched previously.
 var keyRefetchPeriod = 30 * time.Second
 
+// chainStartPollingInterval client will poll chain start event in case beacon chain is
+// offline every at this interval.
```
Comment (rkapka) [inline validator/client/validator.go:48]:
```suggestion
// offline at this interval.
```

**Review by rkapka — COMMENTED** [review]
.

**Review by nisdas — CHANGES_REQUESTED** [review]
The linked issue references 2 different things to resolve.
1) If a beacon node is offline, the validator will permanently be in a crash loop waiting for it to come back up.
2) In the event the beacon node goes offline while the validator is 'waiting' for chainstart, we get a crash again.

The PR currently only tries to address the latter case and doesn't address the former case. The positioning of the liveness check is off over here. If we get any error from the stream we will end up still shutting the node down still. The solution only handles errors in stream creation. If we do succesfully initiate a stream with a beacon node and the node goes down, the validator will still crash.

File: validator/client/validator_test.go [hunk 3]
```diff
@@ -258,17 +259,54 @@ func TestWaitForChainStart_StreamSetupFails(t *testing.T) {
...
+	clientStream := mock.NewMockBeaconNodeValidator_WaitForChainStartClient(ctrl)
+	client.EXPECT().WaitForChainStart(
+		gomock.Any(),
+		&ptypes.Empty{},
+	).Times(11).Return(clientStream, errors.New("failed stream"))
 
+	go func() {
+		err = v.WaitForChainStart(ctx)
+	}()
+	time.Sleep(chainStartPollingInterval * 10)
```
Comment (terencechain) [inline validator/client/validator_test.go:300]: Any way to use channels for syntonization instead of sleep here?

Comment (shayzluf) [reply validator/client/validator_test.go:300]: i am redesigning it today to make it fail safe as @nisdas mentioned
will have a poc today

**Comment by shayzluf** [comment]
replaced by #8278

**Pull request closed without merging**